add draw_net

faizankshaikh opened the pull request:
Added feature for drawing neural network architecture (referred [this](https://gist.github.com/ebenolson/1682625dc9823e27d771) code). Test implementation [here](https://github.com/faizankshaikh/kaggle_right_whale/blob/master/project/whale_classifier.ipynb) in block 10 (draw_to_notebook function)

PS. This is my first contribution, so please feel free to correct my noob mistakes (if any :wink: )

To do:

- [x] How to do a sucessful travis check
- [x] complementary coloring

BenjaminBossan commented:
Hi, this looks pretty cool, we should definitely have something like this. I will have a closer look at the code but here are some questions that popped up:

* Could you please add tests, simple ones like the others for `visualize.py`?
* Don't we need `pydot` as a new dependency?
* Does your code produce good results when the network architecture is not linear (e.g. multiple inputs)?
* I see that you pick a random color for other layer types via hashing. I'm afraid that this may lead to bad colors (like black, so that the font would be illegible). My proposal would be to choose like 20 good looking colors and map onto those (there are online tools that help you choose nice, complementing colors).

Anyway, I believe this would be a welcome addition.

faizankshaikh commented:
1. I would add the tests but the issue is that unlike other functions in visualize (which return a plotted frame) the functions draw_to_file returns a file with specified file format (eg .pdf) and draw_to_notebook returns an IPython Image. So how should I proceed? Also if possible, could you give links to a resources for python testing? (basics and stuff)

2. Yes this needs pydot as a dependency. It also requires Graphviz for its internal working.

3. This gives a good results for non-linear graph too (see [block 19](https://github.com/faizankshaikh/kaggle_right_whale/blob/master/project/whale_localizer.ipynb))

4. Got it. will work on that.

I have some doubts
* Above there's "Travis CI build failed". What does that mean?
* In this PR, the nonlinearity layer is drawn as a completely new layer. I particularly liked [this implementation]() (see page 2 model of the paper) where nonlinearity layer is a sub-layer and is smaller. This appears more cleaner and better. But its out of my horizon of coding :disappointed: 

One more thing, I'll probably be slow (give or take an hour a day), so please bear with me. 

Thanks for hearing me out!

BenjaminBossan commented:
Hi,

1. Visualizations are not that easy to test; for your functions, it would be sufficient if you show that they work with different settings (i.e. don't raise errors). You could use some of the pre-defined nets, as is done [here](https://github.com/dnouri/nolearn/blob/master/nolearn/lasagne/tests/test_visualize.py).

2. Try to just add pydot to the requirements.txt, this should already fix the travis errors. (Travis is a tool that automatically creates an environment and tests your package.)

3. This looks good.

4. Great.

I believe you pasted the wrong link with regard to the nonlinearity layer. Anyway, I would not worry too much about it, there is no harm in fixing this at a later point in time.

Take your time. If you get stuck, upload your code and we may be able to help you out.

dnouri commented:
For 2.: Should also add it to `setup.py`.

Also interested to hear about: "Does your code produce good results when the network architecture is not linear (e.g. multiple inputs)?"

faizankshaikh commented:
Sorry for that... i've edited the link. Will reply for the rest as soon as
I can.

faizankshaikh commented:
Added tests today, still have to work on them but. 

Is there any way to do visualize.py testing locally (what I found that you have to run "$ py.test" but it runs all the tests, instead I only want to run the tests which I created)

There's one problem, there's a python-3 compatibility error. I am still working on that..will reply on how to solve this.

dnouri commented:
You can run

py.test -k test_visualize

to only run tests within that test module. -k also works with test
functions or classes.

BenjaminBossan commented:
I see that for the tests to pass, we would, in addition to adding graphvis, also need to add IPython. I am a little bit concerned that nolearn might have too many dependencies. @dnouri, what about the possibility to make the addition of the visualization tools optional? They have many dependencies now (and matplotlib/pydot sometimes cause trouble) and many people might not care for visualizations.

faizankshaikh commented:
I agree that draw_to_notebook is a bit of an overkill, but you can see that it does nothing special. It's just that people (who like me depend on ipython notebook for literally everything) could gain from it.

But I must ask you to keep the visualizations. They are very helpful and intuitive. Moreover the other good libraries (eg caffe, tensorflow) have their own vis tools. It would be great if nolearn could have this. What we could do is refer the implementations of these libraries and incorporate them here.

I am sorry if there are mishaps on my end. I don't have the expertise or resources to do it as optimally as I can. If you could just give me this chance, I swear I would not let you down.

dnouri commented:
> I am sorry if there are mishaps on my end.

Don't worry.  Nothing bad happened.  Happy to have your contributions.

> What about the possibility to make the addition of the visualization tools optional?

Yes, let's make these visualization dependencies an [extra](https://pythonhosted.org/setuptools/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies).  This way people can install `nolearn[viz]` if they want visualization dependencies.  I'll try to work on this tomorrow.  For now, please feel free to add all the requirements necessary for this PR into the normal `install_requires` to see if all tests can pass this way.  (Or fine if you can tell me they work locally.)

faizankshaikh commented:
Tests run sucessfully on local machine
*  [log.txt](https://github.com/dnouri/nolearn/files/192305/log.txt) 
*  [network.pdf.zip](https://github.com/dnouri/nolearn/files/192308/network.pdf.zip)


#TODO
*  How to do a sucessful travis check
*  complementary coloring

dnouri commented on nolearn/lasagne/tests/test_visualize.py:
```diff
@@ -55,3 +55,15 @@ def test_plot_occlusion_colored_non_square(
+        from nolearn.lasagne.visualize import draw_to_file
+        
+        draw_to_file(net_fitted.get_all_layers(), 'network.pdf', output_shape=False)
+        os.remove('network.pdf')
```
It's maybe better to use pytest's [`tmpdir` fixture](https://pytest.org/latest/tmpdir.html) to create the temporary file.  Something like this should work (untested):

```python
    def test_draw_to_file(self, net_fitted, tmpdir):
        from nolearn.lasagne.visualize import draw_to_file
        fn = str(tmpdir.join('network.pdf'))
        draw_to_file(net_fitted.get_all_layers(), fn, output_shape=False)
```

dnouri commented:
@faizankshaikh Would you please rebase this on top of master?  This way you'll get the `visualization` `extras_require`, which will allow you to solve your IPython dependency issue.  Also some pep8 checking.

faizankshaikh commented:
I get this error now, 

>faizy@faizy-PC ~/workspace/nolearn_contri $ py.test -k test_visualize
usage: py.test [options] [file_or_dir] [file_or_dir] [...]
py.test: error: unrecognized arguments: --flakes --pep8

dnouri commented:
Sorry about that. Test dependencies have changed. Can you do

python setup.py dev

This should install pytest-pep8 and pytest-flakes. Once you're able to run
tests again, you'll notice that there's a few pep8 errors. Which is why I
added pytest-pep8. Thanks!

faizankshaikh commented:
Thanks for the reply. I noticed the dependency error and installed them
manually. That is why I deleted the comment, in hopes that it won't disturb
you (guess I was wrong) And yes, there were pep8 errors...cleaned them.
Thanks again.
On 31-Mar-2016 6:53 pm, "Daniel Nouri" <notifications@github.com> wrote:

> Sorry about that. Test dependencies have changed. Can you do
>
> python setup.py dev
>
> This should install pytest-pep8 and pytest-flakes. Once you're able to run
> tests again, you'll notice that there's a few pep8 errors. Which is why I
> added pytest-pep8. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/dnouri/nolearn/pull/231#issuecomment-203934142>
>

dnouri commented on setup.py:
```diff
@@ -17,6 +17,7 @@
     'scikit-learn',
     'tabulate',
     'Lasagne',
+    'pydot'
```
You can put this and IPython into the `visualization_require` list below.  This way both should get installed when running tests via Travis.  (That's because Travis installs all extras.)

dnouri commented on requirements.txt:
```diff
@@ -5,3 +5,5 @@ Theano==0.8
 scikit-learn==0.17
 tabulate==0.7.5
+graphviz
+pydot
```
The `requirements.txt` currently excludes requirements for visualization.  So safe to remove these two for now.  They'll still be installed via the `visualization_require` list defined in `setup.py`.

dnouri commented:
Try adding `graphviz` to the [list of packages to install on the Travis box](https://github.com/dnouri/nolearn/blob/master/.travis.yml#L11).

dnouri commented on nolearn/lasagne/visualize.py:
```diff
@@ -259,3 +260,106 @@ def saliency_map_net(net, X):
+    """
+    return '#{0:x}'.format(hash(layer_type) % 2**24)
+    
+def get_pydot_graph(layers, output_shape = True, verbose = False):
```
Do you mind calling this `make_pydot_graph` rather than `get_pydot_graph`?

faizankshaikh replied:
Done

faizankshaikh commented:
YES!!!

faizankshaikh commented:
Now have to work on the original problem :smile:

faizankshaikh commented:
Any thoughts on the following would be welcome!

I originally thought (and as @BenjaminBossan suggested) of hardcoding block color and font color. I would suggest that instead of if-else ladder, we could have a data structure (dictionary maybe?) with some complementary colors (I found [this site](http://www.colorhexa.com) helpful) with (random) runtime allocation. Any tips on this?

dnouri commented:
Let's add a very prominent comment saying where we lifted the code from.  (Which you did mention in your original comment, but it's not mentioned in the code itself yet.)

I think I'd also be fine to retain the [original `get_hex_color` implementation](https://gist.github.com/ebenolson/1682625dc9823e27d771#file-draw_net-py-L21) for now.  Unless @BenjaminBossan has a better idea.

dnouri commented on nolearn/lasagne/visualize.py:
```diff
@@ -259,3 +260,110 @@ def saliency_map_net(net, X):
 
 def plot_saliency(net, X, figsize=(9, None)):
     return _plot_heat_map(net, X, figsize, lambda net, X, n: -saliency_map_net(net, X))
+    
```
I'm surprised this doesn't give you a pep8 error, but what we're doing right now is we put two empty lines between functions.  Thanks.

BenjaminBossan replied:
Is it possible that the 2-line requirements stems from pylint, not from pep8?

BenjaminBossan commented:
Let me see if I find a better solution for the colors this weekend.

faizankshaikh commented:
Today I checked with pylint manually and corrected (many!) errors.

BenjaminBossan commented:
I have a proposal for the color picker. I have chosen 4 types of layers (conv, pool, recurrent, other) with each 5 similar colors. Here is the code:

```
COLORS = ['#4A88B3', '#98C1DE', '#6CA2C8', '#3173A2', '#17649B',
          '#FFBB60', '#FFDAA9', '#FFC981', '#FCAC41', '#F29416',
          '#C54AAA', '#E698D4', '#D56CBE', '#B72F99', '#B0108D',
          '#75DF54', '#B3F1A0', '#91E875', '#5DD637', '#3FCD12']


def get_hex_color(layer_type):
    black = '#000000'
    hashed = int(hash(layer_type)) % 5

    if "conv" in layer_type.lower():
        return COLORS[:5][hashed], black
    if layer_type in lasagne.layers.pool.__all__:
        return COLORS[5:10][hashed], black
    if layer_type in lasagne.layers.recurrent.__all__:
        return COLORS[10:15][hashed], black
    else:
        return COLORS[15:20][hashed], black
```

On my display at least, the colors look fine. There are some colors that are never used with the current layers (`'#3173A2', '#4A88B3', '#98C1DE', '#B72F99', '#C54AAA', '#D56CBE'`) but that may of course change with the addition of more layers. Maybe you have an idea how to hash onto the colors more evenly.

I also have a proposal: Could you change the code so that `make_pydot_graph`, `draw_to_file`, and `draw_to_notebook` would also accept a `NeuralNet` instance as first argument? The reason is that none of the other functions accept a list of layers as argument; I could imagine that many nolearn users would find passing the net as argument to be more intuitive.

dnouri commented:
@BenjaminBossan I like the solution to the color picking.  Two things though:
- Can we drop though the second argument we're returning (the black in all cases); that's not used in the original implementation.
- Maybe make `COLORS` a 2d list, this way we can get rid of the slicing.

> Could you change the code so that make_pydot_graph, draw_to_file, and draw_to_notebook would also accept a NeuralNet instance as first argument?

Agree that would be useful.

@BenjaminBossan Can we following the merge of this update your notebook to include the rendered network in one of the examples?

faizankshaikh commented:
@dnouri : The second argument is necessary to change the font color (not in the original implementation, I tweaked it a bit)

dnouri commented:
> The second argument is necessary to change the font color (not in the original implementation, I tweaked it a bit)

OK, do we need it still though?  Seems like font color could always be black?

faizankshaikh commented:
As per @BenjaminBossan above,
> I'm afraid that this may lead to bad colors (like black, so that the font would be illegible)

was a valid point

BenjaminBossan commented:
I agree that we should drop the second color, makes it simpler. With the current colors, black should work. If we ever need other font colors, we could add a similar function that turns layer names to font colors.

> Maybe make COLORS a 2d list, this way we can get rid of the slicing.

I thought about a dictionary but then if we ever need the colors for something else, a list might be easier and slicing does not add too much overhead -- but I don't care.

> Can we following the merge of this update your notebook to include the rendered network in one of the examples?

I will update it soon, some other parts are not up-to-date either.

faizankshaikh commented:
Alright then. I'll cook the PR by tomorrow. Thanks for your support @BenjaminBossan @dnouri

faizankshaikh commented:
[Done](http://nbviewer.jupyter.org/gist/faizankshaikh/f93d2182135ef975be160076bb9070b6)!

dnouri commented on nolearn/lasagne/visualize.py:
```diff
@@ -338,29 +347,31 @@ def make_pydot_graph(layers, output_shape=True, verbose=False):
 
 
-def draw_to_file(layers, filename, **kwargs):
+def draw_to_file(net, filename, **kwargs):
```
Let's try to keep backward compatibility and make both layers and net work as an argument.  So maybe keep calling the argument `layers` and further down do: `layers = layers.get_all_layers() if hasattr(layers, 'get_all_layers') else layers`

faizankshaikh replied:
May I ask why to do that? 

Let me make an argument. We pass the "net" for abstracting the implementation to the user and make the function more intuitive right? So why pass the layers?

BenjaminBossan replied:
I'm also not sure why we would need backwards compatibility here.

Anyway, if both modes were supported, could you please adjust the tests to test both?

dnouri replied:
To keep backward compatibility with the code that we started out with.  If a Lasagne user wants to use Eben's code packaged up in a library, they now can.

dnouri commented on nolearn/lasagne/visualize.py:
```diff
@@ -270,14 +271,23 @@ def get_hex_color(layer_type):
     """
-    if 'Conv' in layer_type:
-        return '#ff0000', 'cyan'
+    COLORS = ['#4A88B3', '#98C1DE', '#6CA2C8', '#3173A2', '#17649B', \
```
The backslashes at the end here are unnecessary

faizankshaikh replied:
There was a pylint error there, that is why I added that

BenjaminBossan replied:
That seems odd, could you tell which error pylint finds?

dnouri replied:
Hmm, the pylint is probably asking you to align the continued lines below differently.  The `\` should go in any case.

faizankshaikh replied:
It says
> C:277, 0: Line too long (121/100) (line-too-long)

faizankshaikh replied:
I am sorry I did not understand you :disappointed:

dnouri replied:
Write:

```python
    COLORS = ['#4A88B3', '#98C1DE', '#6CA2C8', '#3173A2', '#17649B',
              '#FFBB60', '#FFDAA9', '#FFC981', '#FCAC41', '#F29416',
```

See how the two lines align?

faizankshaikh replied:
It still gives the same error
>C:277, 0: Wrong continued indentation.
                       '#FFBB60', '#FFDAA9', '#FFC981', '#FCAC41', '#F29416',
              |        ^ (bad-continuation)

faizankshaikh commented on nolearn/lasagne/visualize.py:
```diff
@@ -338,29 +347,31 @@ def make_pydot_graph(layers, output_shape=True, verbose=False):
 
 
-def draw_to_notebook(layers, **kwargs):
+def draw_to_notebook(net, **kwargs):
```
One thing I'm afraid of here is when (unknowingly) the user passes a net object which is not initialized. How could you handle this?

BenjaminBossan replied:
You could check whether the net is initialized and if not tell the user so?

dnouri commented on nolearn/lasagne/visualize.py:
```diff
@@ -259,3 +262,116 @@ def saliency_map_net(net, X):
+        - color : string containing a hex color for filling block.
+    """
+    COLORS = ['#4A88B3', '#98C1DE', '#6CA2C8', '#3173A2', '#17649B', \
+          '#FFBB60', '#FFDAA9', '#FFC981', '#FCAC41', '#F29416', \
```
Looks like you're missing four space characters at the start of this line.  Same for the following two lines.  Please remove the `\` at the end.

faizankshaikh replied:
Ok I'm removing.

faizankshaikh commented:
Is there anything left for me to do?

BenjaminBossan commented:
Be patient, I'm sure Daniel will have another look at it and merge it when he believes it's ready.

dnouri commented:
Thanks @faizankshaikh
